torch/nn/functional.py

```diff
@@ -3075,90 +3075,98 @@ def interpolate(input, size=None, scale_factor=None, mode='nearest', align_corne
             output_size = size
         else:
-            output_size = [size for i in range(dim)]
-    else:
-        assert scale_factor is not None
+            output_size = [size for _ in range(dim)]
+    elif scale_factor is not None:
+        assert size is None
+        output_size = None
         if isinstance(scale_factor, (list, tuple)):
+            if len(scale_factor) != dim:
+                raise ValueError('scale_factor shape must match input shape. '
+                                 'Input is {}D, scale_factor is {}'.format(dim, len(scale_factor)))
             scale_factors = scale_factor
         else:
             scale_factors = [scale_factor for _ in range(dim)]
+    else:
+        raise ValueError('either size or scale_factor should be defined')
 
-        if recompute_scale_factor is None:
-            # only warn when the scales have floating values since
-            # the result for ints is the same with/without recompute_scale_factor
-
-            is_float_scale_factor = False
+    if recompute_scale_factor is None:
+        # only warn when the scales have floating values since
+        # the result for ints is the same with/without recompute_scale_factor
+        if scale_factors is not None:
             for scale in scale_factors:
-                is_float_scale_factor = math.floor(scale) != scale
-                if is_float_scale_factor:
+                if math.floor(scale) != scale:
+                    warnings.warn("The default behavior for interpolate/upsample with float scale_factor changed "
+                                  "in 1.6.0 to align with other frameworks/libraries, and now uses scale_factor directly, "
+                                  "instead of relying on the computed output size. "
+                                  "If you wish to restore the old behavior, please set recompute_scale_factor=True. "
+                                  "See the documentation of nn.Upsample for details. ")
                     break
+    elif recompute_scale_factor and size is not None:
+        raise ValueError("recompute_scale_factor is not meaningful with an explicit size.")
 
-            if is_float_scale_factor:
-                warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
-                              "in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, "
-                              "instead of relying on the computed output size. "
-                              "If you wish to keep the old behavior, please set recompute_scale_factor=True. "
-                              "See the documentation of nn.Upsample for details. ")
+    # "area" mode always requires an explicit size rather than scale factor.
+    # Re-use the recompute_scale_factor code path.
+    if mode == "area" and output_size is None:
+        recompute_scale_factor = True
 
+    if recompute_scale_factor is not None and recompute_scale_factor:
```

Comment:
isn't this the same as `if recompute_scale_factor`?

Reply:
Yeah, but that doesn't work in TorchScript.

Reply:
what a time to be alive.